Classes/Routing/FrontendNodeRoutePartHandler.php

```diff
@@ -78,7 +78,34 @@ protected function getRelativeNodePathByUriPathSegmentProperties(NodeInterface $
             return NodePaths::getRelativePathBetween($siteNode->getPath(), $foundNode->getPath());
         }
 
-        return parent::getRelativeNodePathByUriPathSegmentProperties($siteNode, $relativeRequestPath);
+        // Hotfix for performance issue in original Neos route part handler: if a given node contains thousands of
+        // child nodes, the original implementation would load all these nodes into memory in order to compare the
+        // current path segment with $node->getProperty('uriPathSegment').
+        //
+        // When the issue is fixed in Neos core, the following code can be removed by a parent::… call.
+        $relativeNodePathSegments = [];
+        $node = $siteNode;
+
+        foreach (explode('/', $relativeRequestPath) as $pathSegment) {
+            $foundNodeInThisSegment = false;
+
+            $possibleNodes = $this->nodeDataRepository->findByProperties(['uriPathSegment' => $pathSegment], 'Neos.Neos:Document', $siteNode->getWorkspace(), $siteNode->getDimensions(), $node->getPath());
```

Comment:
nodeDataRepository needs to be injected. Apart from that seems good!

Reply:
nodeDataRepository is already present in the parent classe …

Reply:
Only in your hacked version ;)

https://github.com/neos/neos-development-collection/blob/master/Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php

Reply:
That's true. I only wanted to know if you take a close look ☺️